File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/25901f324550ce931411ed9a1ed24a37.py
class Sparsetable_min:
    def __init__(self):
        self.beki = 19
        self.cor = 1 << self.beki
        self.vec = [[0] * self.cor for _ in range(self.beki + 1)]
        self.lo = [-1] + [0] * self.cor
        for i in range(1, self.cor + 1):
            self.lo[i] = self.lo[i >> 1] + 1
    
    def shoki1(self):
        self.vec = [[0] * self.cor for _ in range(self.beki + 1)]
        self.lo = [-1] + [0] * self.cor
        for i in range(1, self.cor + 1):
            self.lo[i] = self.lo[i >> 1] + 1
    
    def shoki2(self):
        for i in range(1, self.beki + 1):
            for j in range(self.cor):
                if j + (1 << (i - 1)) < self.cor:
                    self.vec[i][j] = min(self.vec[i - 1][j], self.vec[i - 1][j + (1 << (i - 1))])
                else:
                    self.vec[i][j] = self.vec[i - 1][j]
    
    def rangemin(self, i, j):
        haba = self.lo[j - i]
        return min(self.vec[haba][i], self.vec[haba][j - (1 << haba)])


class SAIS:
    def __init__(self):
        self.STR = ""
        self.S = []
        self.SL = []
        self.haba = []
        self.SA = []
        self.rank = []
        self.LCP = []
        self.mojisu = 0
        self.len = 0
    
    def make_suffix_array_string(self, string):
        self.STR = string
        setc = set(string)
        mapc = {}
        cnt = 1
        for c in setc:
            mapc[c] = cnt
            cnt += 1
        self.S = [mapc[c] for c in string]
        self.mojisu = cnt
        self.S.append(0)
        self.make_suffix_array_naibu()
    
    def make_suffix_array_saiki(self, ve, max_valu):
        self.S = ve
        self.S.append(0)
        self.mojisu = max_valu + 1
        self.make_suffix_array_naibu()
    
    def make_suffix_array_naibu(self):
        cnt = self.mojisu
        self.len = len(self.S)
        kosuu = [0] * cnt
        for c in self.S:
            kosuu[c] += 1
        
        self.haba.append((0, 0))
        for i in range(1, cnt):
            self.haba.append((self.haba[-1][1] + 1, self.haba[-1][1] + kosuu[i]))
        
        self.SL = [0] * self.len
        for i in range(self.len - 1, -1, -1):
            if i == self.len - 1:
                self.SL[i] = 1
            else:
                if self.S[i] < self.S[i + 1]:
                    self.SL[i] = 1
                elif self.S[i] > self.S[i + 1]:
                    self.SL[i] = 0
                else:
                    self.SL[i] = self.SL[i + 1]
        
        LMS_num = 0
        LMS = []
        LMS_id = [0] * self.len
        rt = 0
        for i in range(1, self.len):
            if self.SL[i - 1] == 0 and self.SL[i] == 1:
                LMS.append(i)
                LMS_id[i] = rt
                LMS_num += 1
                rt += 1
        
        for i in range(LMS_num - 1):
            self.SL[LMS[i]] = -(LMS[i + 1] - LMS[i] + 1)
        self.SL[self.len - 1] = -1
        
        self.SA = [-1] * self.len
        pos = [0] * cnt
        for c in range(self.len):
            if self.SL[c] < 0:
                self.SA[self.haba[self.S[c]][1] - pos[self.S[c]]] = c
                pos[self.S[c]] += 1
        
        self.induced_sort(self.SA, self.len, cnt)
        
        lms_sub_order = []
        mae_ind = 0
        for i in range(self.len):
            if self.SL[self.SA[i]] < 0:
                if i == 0:
                    lms_sub_order.append((self.SA[i], 1))
                elif lms_sub_order[-1][1] == 1:
                    lms_sub_order.append((self.SA[i], 2))
                elif self.SL[self.SA[mae_ind]] != self.SL[self.SA[i]]:
                    lms_sub_order.append((self.SA[i], lms_sub_order[-1][1] + 1))
                else:
                    lmslen = -self.SL[self.SA[mae_ind]]
                    bo = True
                    for j in range(lmslen):
                        if self.S[self.SA[mae_ind] + j] != self.S[self.SA[i] + j]:
                            bo = False
                            break
                    if bo:
                        lms_sub_order.append((self.SA[i], lms_sub_order[-1][1]))
                    else:
                        lms_sub_order.append((self.SA[i], lms_sub_order[-1][1] + 1))
                mae_ind = i
        
        rec = [0] * len(LMS)
        for i in range(LMS_num):
            rec[LMS_id[lms_sub_order[i][0]]] = lms_sub_order[i][1]
        
        ss_naibu = SAIS()
        ss_naibu.make_suffix_array_saiki(rec, lms_sub_order[-1][1])
        
        for i in range(LMS_num, 0, -1):
            c = LMS[ss_naibu.SA[i]]
            self.SA[self.haba[self.S[c]][1] - pos[self.S[c]]] = c
            pos[self.S[c]] += 1
        
        self.induced_sort(self.SA, self.len, cnt)
        self.rank = [0] * self.len
        for i in range(self.len):
            self.rank[self.SA[i]] = i
    
    def induced_sort(self, ve, length, cnt):
        ue = [0] * cnt
        for i in range(length):
            if ve[i] == -1:
                continue
            if ve[i] == 0:
                continue
            if self.SL[ve[i] - 1] == 0:
                ve[self.haba[self.S[ve[i] - 1]][0] + ue[self.S[ve[i] - 1]]] = ve[i] - 1
                ue[self.S[ve[i] - 1]] += 1
            if self.SL[ve[i]] < 0 and i > 0:
                ve[i] = -1
        ue = [0] * cnt
        for i in range(length - 1, -1, -1):
            if ve[i] == -1:
                continue
            if ve[i] == 0:
                continue
            if self.SL[ve[i] - 1] != 0:
                ve[self.haba[self.S[ve[i] - 1]][1] - ue[self.S[ve[i] - 1]]] = ve[i] - 1
                ue[self.S[ve[i] - 1]] += 1
        for v in ve:
            if v == -1:
                assert False
    
    def make_lcp(self):
        self.LCP = [0] * self.len
        h = 0
        self.LCP[0] = 0
        for i in range(self.len - 1):
            j = self.SA[self.rank[i] - 1]
            if h > 0:
                h -= 1
            while j + h < self.len - 1 and i + h < self.len - 1:
                if self.S[j + h] != self.S[i + h]:
                    break
                h += 1
            self.LCP[self.rank[i] - 1] = h
    
    def lower_bound(self, p):
        m = len(p)
        ue, si = self.len, 0
        while ue - si > 1:
            me = (ue + si) // 2
            if self.STR.compare(self.SA[me], m, p) < 0:
                si = me
            else:
                ue = me
        return ue
    
    def exist(self, p):
        l = len(p)
        id = self.lower_bound(p)
        return self.STR.compare(self.SA[id], l, p) == 0
    
    def range(self, p):
        return (self.lower_bound(p), self.lower_bound(p + '{'))


def main():
    n = int(input())
    s = input()
    ve = [0] * n
    cnt = 0
    t = ""
    for i in range(n):
        if s[i] == '1':
            ve[i] = cnt
        else:
            ve[i] = cnt
            cnt += 1
            if i % 2 == 1:
                t += 'b'
            else:
                t += 'a'
    m = len(t)
    if m == 0:
        e = int(input())
        for _ in range(e):
            print("Yes")
        return
    for i in range(m):
        t += chr(ord('a') ^ ord('b') ^ ord(t[i]))
    
    ve.append(cnt)
    ss = SAIS()
    ss.make_suffix_array_string(t)
    ss.make_lcp()
    sm = Sparsetable_min()
    sm.shoki1()
    for i in range(len(ss.LCP)):
        sm.vec[0][i] = ss.LCP[i]
    sm.shoki2()
    q = int(input())
    for _ in range(q):
        l1, l2, p = map(int, input().split())
        l1 -= 1
        l2 -= 1
        r1 = l1 + p
        r2 = l2 + p
        if ve[r1] - ve[l1] != ve[r2] - ve[l2]:
            print("No")
            continue
        if l1 % 2 == l2 % 2:
            if ve[r1] == ve[l1] or ve[r1] == ve[r2]:
                print("Yes")
                continue
            e1 = ss.rank[ve[l1]]
            e2 = ss.rank[ve[l2]]
            if e1 > e2:
                e1, e2 = e2, e1
            if sm.rangemin(e1, e2) >= ve[r1] - ve[l1]:
                print("Yes")
            else:
                print("No")
        else:
            if ve[r1] == ve[l1]:
                print("Yes")
                continue
            if ve[r1] == ve[r2]:
                print("No")
                continue
            e1 = ss.rank[ve[l1]]
            e2 = ss.rank[ve[l2] + m]
            if e1 > e2:
                e1, e2 = e2, e1
            if sm.rangemin(e1, e2) >= ve[r1] - ve[l1]:
                print("Yes")
            else:
                print("No")


if __name__ == "__main__":
    main()

File: dataset/xCodeEval_4000/C++/Code/25901f324550ce931411ed9a1ed24a37.cpp
   #include <bits/stdc++.h>
 




 #define inf  1000000007
 #define pa pair<int,int>
 #define ll long long
 #define pal pair<double,double>
 #define ppap pair<pa,int>
  #define PI 3.14159265358979323846
  #define paa pair<int,char>
  #define  mp make_pair
  #define  pb push_back
  #define EPS (1e-8)
                                          
    int dx[8]={0,1,0,-1,1,1,-1,-1};
    int dy[8]={1,0,-1,0,-1,1,1,-1};
                                            using namespace std;
                                   			class pa3{
                                            	public:
                                            	int x;
                                   				int y,z;
                                            	pa3(int x=0,int y=0,int z=0):x(x),y(y),z(z) {}
                                            	bool operator < (const pa3 &p) const{
                                            		if(x!=p.x) return x<p.x;
                                            		if(y!=p.y) return y<p.y;
                                            		 return z<p.z;
                                            		

                                            	}
                                   				bool operator > (const pa3 &p) const{
                                            		if(x!=p.x) return x>p.x;
                                            		if(y!=p.y) return y>p.y;
                                            		 return z>p.z;
                                            		

                                            	}
                                            	bool operator == (const pa3 &p) const{
                                            		return x==p.x && y==p.y && z==p.z;
                                            	}
                                            		bool operator != (const pa3 &p) const{
                                            			return !( x==p.x && y==p.y && z==p.z);
                                            	}
                                            
                                            };
                                            
                                            class pa4{
                                            	public:
                                            	int x;
                                            	int y,z,w;
                                            	pa4(int x=0,int y=0,int z=0,int w=0):x(x),y(y),z(z),w(w) {}
                                            	bool operator < (const pa4 &p) const{
                                            		if(x!=p.x) return x<p.x;
                                            		if(y!=p.y) return y<p.y;
                                            		if(z!=p.z)return z<p.z;
                                            		return w<p.w;
                                            		

                                            	}
                                            	bool operator > (const pa4 &p) const{
                                            		if(x!=p.x) return x>p.x;
                                            		if(y!=p.y) return y>p.y;
                                            		if(z!=p.z)return z>p.z;
                                            		return w>p.w;
                                            		

                                            	}
                                            	bool operator == (const pa4 &p) const{
                                            		return x==p.x && y==p.y && z==p.z &&w==p.w;
                                            	}
                                            		
                                            
                                            };
                                            class pa2{
                                            	public:
                                            	int x,y;
                                            	pa2(int x=0,int y=0):x(x),y(y) {}
                                            	pa2 operator + (pa2 p) {return pa2(x+p.x,y+p.y);}
                                            	pa2 operator - (pa2 p) {return pa2(x-p.x,y-p.y);}
                                            	bool operator < (const pa2 &p) const{
                                            		return y != p.y ? y<p.y: x<p.x;
                                            	}
                                            	bool operator > (const pa2 &p) const{
                                            		return x != p.x ? x<p.x: y<p.y;
                                            	}
                                            	bool operator == (const pa2 &p) const{
                                            		return abs(x-p.x)==0 && abs(y-p.y)==0;
                                            	}
                                            	bool operator != (const pa2 &p) const{
                                            		return !(abs(x-p.x)==0 && abs(y-p.y)==0);
                                            	}
                                            		
                                            
                                            };
                                            
 
                      
                                string itos( int i ) {
                                ostringstream s ;
                                s << i ;
                                return s.str() ;
                                }
                                 
                                int gcd(int v,int b){
                                	if(v==0) return b;
                                	if(b==0) return v;
                                	if(v>b) return gcd(b,v);
                                	if(v==b) return b;
                                	if(b%v==0) return v;
                                	return gcd(v,b%v);
                                }
                 
                            
                                int mod;
int extgcd(int a, int b, int &x, int &y) {
    if (b == 0) {
        x = 1;
        y = 0;
        return a;
    }
    int d = extgcd(b, a%b, y, x);
    y -= a/b * x;
    return d;
}
pa operator+(const pa & l,const pa & r) {   
    return {l.first+r.first,l.second+r.second};                                    
}    
pa operator-(const pa & l,const pa & r) {   
    return {l.first-r.first,l.second-r.second};                                    
}  
              
              

    		

    			

                 

 

  


struct Sparsetable_min{
	
	
	private:
	public:
	
	

	

	

	

	
	int beki=19;
	
	int cor=(1<<beki);
	
	vector<vector<int>> vec;
	vector<int> lo;
	
	void shoki1(){
		vec= vector<vector<int>>(beki+1, vector<int>(cor, 0));
		lo=vector<int>(cor+1);
		lo[0]=-1;
		for(int i=1;i<=cor;i++)lo[i]=lo[i>>1]+1;
	}
	
	void shoki2(){
		for(int i=1;i<=beki;i++)for(int j=0;j<cor;j++){
			if((j+(1<<(i-1)))<cor)vec[i][j]=min(vec[i-1][j],vec[i-1][j+(1<<(i-1))]);
			else vec[i][j]=vec[i-1][j];
		}
	}
	
	int rangemin(int i,int j){
		int haba=lo[j-i];
		
		return min(vec[haba][i],vec[haba][j-(1<<haba)]);
	}
	
};


struct SAIS{
	private:
	public:

	string STR;
	vector<int> S;
	vector<int> SL;
	

	

	

	vector<pa> haba;
	vector<int> SA;
	vector<int> rank;
	vector<int> LCP;
	int mojisu;
	int len;
	
	void make_suffix_array_string(string str){
		STR=str;
		set<char> setc;
		map<char,int> mapc;
		for(auto c:str)setc.insert(c);
		int cnt=1;
		for(auto it=setc.begin();it!=setc.end();it++){
			mapc[*it]=cnt;
			cnt++;
		}
		for(auto c:str)S.pb(mapc[c]);
		mojisu=cnt;
		S.pb(0);
		make_suffix_array_naibu();
	}
	
	void make_suffix_array_saiki(vector<int> ve,int max_valu){
		S=ve;
		S.pb(0);
		mojisu=max_valu+1;
		make_suffix_array_naibu();
	}
	
	
	void make_suffix_array_naibu(){
		int cnt=mojisu;
		len=S.size();
		
		vector<int> kosuu(cnt,0);
		for(auto c:S)kosuu[c]++;
		
		haba.pb({0,0});
		for(int i=1;i<cnt;i++){
			haba.pb({haba.back().second+1,haba.back().second+kosuu[i]});
		}
		
		

		

		

		

		

		


		SL.resize(len);
		for(int i=len-1;i>=0;i--){
			if(i==len-1){
				SL[i]=1;
			}
			else{
				if(S[i]<S[i+1]) SL[i]=1;
				else if(S[i]>S[i+1])SL[i]=0;
				else{
					SL[i]=SL[i+1];
				}
			}
		}
		
		int LMS_num=0;
		vector<int> LMS;
		vector<int> LMS_id(len,0);
		int rt=0;
		for(int i=1;i<len;i++)if(SL[i-1]==0 &&SL[i]==1){
			LMS.pb(i);
			LMS_id[i]=rt;
			LMS_num++;
			rt++;
		}
		for(int i=0;i<LMS_num-1;i++){
			SL[LMS[i]]=-(LMS[i+1]-LMS[i]+1);
		}
		SL[len-1]=-1;
		
		for(int i=0;i<len;i++)SA.pb(-1);
		vector<int> pos(cnt,0);
		for(int c=0;c<len;c++)if(SL[c]<0){
			
			SA[haba[S[c]].second-pos[S[c]]]=c;
			pos[S[c]]++;
		}
		

		

		

		

		

		

		
		
		induced_sort(SA,len,cnt);
		
	

		

		

		

		

		

		

		
		vector<pa> lms_sub_order;
		int mae_ind;
		for(int i=0;i<len;i++)if(SL[SA[i]]<0){
			if(i==0){
				lms_sub_order.pb({SA[i],1});
				
			}
			else if(lms_sub_order.back().second==1){
				lms_sub_order.pb({SA[i],2});
			}
			else if(SL[SA[mae_ind]]!=SL[SA[i]]){
				lms_sub_order.pb({SA[i],lms_sub_order.back().second+1});
			}
			else{
				int lmslen= -SL[SA[mae_ind]];
				bool bo=true;
				for(int j=0;j<lmslen;j++){
					if(S[SA[mae_ind]+j]!=S[SA[i]+j]){
						bo=0;
						break;
					}
				}
				if(bo)lms_sub_order.pb({SA[i],lms_sub_order.back().second});
				else lms_sub_order.pb({SA[i],lms_sub_order.back().second+1});
			}
			mae_ind=i;
		}
	

	

			for(int i=0;i<cnt;i++)pos[i]=0;
			for(int i=0;i<len;i++)SA[i]=-1;
		
		if(LMS_num==lms_sub_order.back().second){
			for(int i=LMS_num-1;i>=0;i--){
				int c=lms_sub_order[i].first;
				SA[haba[S[c]].second-pos[S[c]]]=c;
				pos[S[c]]++;
			}
	
		}
		else{
			vector<int> rec((int)LMS.size());
			for(int i=0;i<LMS_num;i++){
				rec[LMS_id[lms_sub_order[i].first]]=lms_sub_order[i].second;
			}
			SAIS ss_naibu;
			ss_naibu.make_suffix_array_saiki(rec,lms_sub_order.back().second);
			
			for(int i=LMS_num;i>0;i--){
				int c=LMS[ss_naibu.SA[i]];
				SA[haba[S[c]].second-pos[S[c]]]=c;
				pos[S[c]]++;
			}
		}

			induced_sort(SA,len,cnt);
		rank.resize(len);
		for(int i=0;i<len;i++)rank[SA[i]]=i;
			return;
	}
	
	void induced_sort(vector<int> &ve, int len,int cnt){
		vector<int> ue(cnt,0);
		for(int i=0;i<len;i++){
			if(ve[i]==-1) continue;
			if(ve[i]==0) continue;
			if(SL[ve[i]-1]==0){
				ve[haba[S[ve[i]-1]].first+ue[S[ve[i]-1]]]=ve[i]-1;
				ue[S[ve[i]-1]]++;
			}
			if(SL[ve[i]]<0 && i>0) ve[i]=-1;
		}
		for(int i=0;i<cnt;i++)ue[i]=0;
		
		for(int i=len-1;i>=0;i--){
			if(ve[i]==-1) continue;
			if(ve[i]==0) continue;
			if(SL[ve[i]-1]!=0){
				ve[haba[S[ve[i]-1]].second-ue[S[ve[i]-1]]]=ve[i]-1;
				ue[S[ve[i]-1]]++;
			}
		}
		for(auto v:ve)if(v==-1)assert(0);
	}
	
	void make_lcp(){
				S.clear();
		 SL.clear();
 haba.clear();

		LCP.resize(len);
		int h=0;
		LCP[0]=0;
		for(int i=0;i<len-1;i++){
			int j=SA[rank[i]-1];
			if(h>0)h--;
			for(;j+h<len-1&& i+h<len-1;h++){
				if(S[j+h]!=S[i+h])break;
			}
			LCP[rank[i]-1]=h;
		}
	

		
		return;
	}
	
	int lower_bound(string p){

		int m=p.length();
		int ue=len,si=0,me;
		while(ue-si>1){
			me=(ue+si)/2;
			if(STR.compare(SA[me],m,p)<0)si=me;
			else ue=me;
		}
		return ue;
	}
	
	bool exist(string p){
		int l=p.length();
		int id=lower_bound(p);
		
	
		return 0==STR.compare(SA[id],l,p);
	}
	pa range(string p){

				return {lower_bound(p),lower_bound(p+'{')};
	}
};

signed main(){
cin.tie(0);
ios::sync_with_stdio(false);

	int n;
	cin>>n;
string s;
	cin>>s;
	vector<int> ve(n);
	int cnt=0;
	string t="";
	for(int i=0;i<n;i++){
		if(s[i]=='1'){
			ve[i]=cnt;
		}
		else{
			ve[i]=cnt;
			cnt++;
			if(i%2)t+="b";
			else t+="a";
		}
	}
		int m=t.length();
	if(m==0){
		int e;
		cin>>e;
		for(int i=0;i<e;i++)cout<<"Yes\n";
		return 0;
	}
	for(int i=0;i<m;i++)t+=(char)('a'^'b'^t[i]);
	
	
	ve.pb(cnt);
	SAIS ss;



	ss.make_suffix_array_string(t);
	ss.make_lcp();
	Sparsetable_min sm;
	sm.shoki1();
	for(int i=0;i<ss.LCP.size();i++)sm.vec[0][i]=ss.LCP[i];




	sm.shoki2();
	int q;
	cin>>q;
	for(int i=0;i<q;i++){
		int l1,l2,p;
		cin>>l1>>l2>>p;
		l1--,l2--;
		int r1=l1+p,r2=l2+p;
		if(ve[r1]-ve[l1]!=ve[r2]-ve[l2]){
			cout<<"No\n";
			continue;
		}
		if(l1%2==l2%2){
			if(ve[r1]==ve[l1] || ve[r1]==ve[r2]){
				cout<<"Yes\n";
					continue;
			}
			int e1=ss.rank[ve[l1]];
			int e2=ss.rank[ve[l2]];
			if(e1>e2)swap(e1,e2);
			if(sm.rangemin(e1,e2)>=ve[r1]-ve[l1]){
				cout<<"Yes\n";
			}
			else cout<<"No\n";
		}
		else{
			if(ve[r1]==ve[l1]){
				cout<<"Yes\n";
					continue;
			}
			if(ve[r1]==ve[r2]){
				cout<<"No\n";
				continue;
			}
			
			int e1=ss.rank[ve[l1]];
			int e2=ss.rank[ve[l2]+m];
			if(e1>e2)swap(e1,e2);
		

			if(sm.rangemin(e1,e2)>=ve[r1]-ve[l1]){
				cout<<"Yes\n";
			}
			else cout<<"No\n";	
		}
	}
return 0;
 }
